src/vmcs/src/ept_intel_x64.cpp

```diff
@@ -42,36 +42,50 @@ ept_entry_intel_x64
 ept_intel_x64::add_page(integer_pointer gpa, integer_pointer bits, integer_pointer end)
 {
     auto &&index = ept::index(gpa, bits);
+    auto &&view = gsl::make_span(m_ept, ept::num_entries);
 
     if (bits > end)
     {
+        auto &&entry = ept_entry_intel_x64(&view.at(index));
+        if (entry.entry_type())
+            throw std::logic_error("unmap gpa before adding new page");
+
         if (m_epts.empty())
             m_epts = std::vector<std::unique_ptr<ept_intel_x64>>(ept::num_entries);
 
         auto &&iter = bfn::find(m_epts, index);
-        if (!(*iter))
-        {
-            auto &&view = gsl::make_span(m_ept, ept::num_entries);
-            (*iter) = std::make_unique<ept_intel_x64>(&view.at(index));
-        }
+        if (nullptr == *iter)
+            *iter = std::make_unique<ept_intel_x64>(&view.at(index));
 
         return (*iter)->add_page(gpa, bits - ept::pt::size, end);
     }
 
     if (!m_epts.empty())
     {
-        m_epts.clear();
-        m_epts.shrink_to_fit();
+        auto &&iter = bfn::find(m_epts, index);
+        if (nullptr != *iter)
+            throw std::logic_error("unmap gpa before adding new page");
     }
 
-    auto &&view = gsl::make_span(m_ept, ept::num_entries);
+    if (is_bit_set(view.at(index), 7))
+        return ept_entry_intel_x64(&view.at(index));
+
+    view.at(index) = 0x80;
```

Comment:
same thing here, not a big fan of setting a bit like that manually. I would prefer this was done through the entry class.